scipy/spatial/transform/rotation.py

```diff
@@ -8,7 +8,107 @@
+
+def compute_euler_from_dcm(dcm, seq, extrinsic=False):
+    # The algorithm assumes intrinsic frame transformations. For representation
+    # the paper uses transformation matrices, which are transpose of the
+    # direction cosine matrices used by our Rotation class.
+    # Adapt the algorithm for our case by
+    # 1. Transposing our dcm representation
+    # 2. Reversing both axis sequence and angles for extrinsic rotations
+
+    if extrinsic:
+        seq = seq[::-1]
+
+    if dcm.ndim == 2:
+        dcm = dcm[None, :, :]
+    num_rotations = dcm.shape[0]
+
+    # Step 0
+    # Algorithms dcms are transpose of ours
+    dcm = np.transpose(dcm, (0, 2, 1))
+    # Algorithm assumes axes as column vectors, here we use 1D vectors
+    n1 = elementary_basis_vector(seq[0])
+    n2 = elementary_basis_vector(seq[1])
+    n3 = elementary_basis_vector(seq[2])
+
+    # Step 2
+    sl = np.dot(np.cross(n1, n2), n3)
+    cl = np.dot(n1, n3)
+
+    # angle phi is lambda from the paper referenced in [2] from docstring of
+    # `as_euler` function
+    phi = np.arctan2(sl, cl)
+    c = np.empty((3, 3))
+    c[0] = n2
+    c[1] = np.cross(n1, n2)
+    c[2] = n1
+
+    # Step 3
+    rt = np.array([
+        [1, 0, 0],
+        [0, cl, -sl],
+        [0, sl, cl]
+    ])
+    rtc = rt.dot(c)
+    ct = c.T
+    res = np.einsum('...ij,...jk->...ik', rtc, dcm)
+    o = np.einsum('...ij,...jk->...ik', res, ct)
+
+    # Step 4
+    offset = np.arccos(o[:, 2, 2])
+    angle2 = phi + offset
+
+    # Steps 5, 6
+    eps = np.finfo(float).resolution  # ~1e-15
+    safe1 = (np.abs(offset) >= eps)
+    safe2 = (np.abs(offset - np.pi) >= eps)
+
+    angle1 = np.empty(num_rotations)
+    angle3 = np.empty(num_rotations)
+
+    # 5b
+    safe_mask = np.logical_and(safe1, safe2)
+    angle1[safe_mask] = np.arctan2(o[safe_mask, 2, 0], -o[safe_mask, 2, 1])
+    angle3[safe_mask] = np.arctan2(o[safe_mask, 0, 2], o[safe_mask, 1, 2])
+
+    # 6a
+    angle3[~safe_mask] = 0
+    # 6b
+    angle1[~safe1] = np.arctan2(o[~safe1, 0, 1] - o[~safe1, 1, 0],
+                                o[~safe1, 0, 0] + o[~safe1, 1, 1])
+    # 6c
+    angle1[~safe2] = np.arctan2(o[~safe2, 0, 1] + o[~safe2, 1, 0],
+                                o[~safe2, 0, 0] - o[~safe2, 1, 1])
+
+    # Step 7
+    # python modulo operator works correctly for negative numbers
+    if seq[0] == seq[2]:
+        # lambda = 0, so we can only ensure angle2 -> [0, pi]
+        adjust_mask = np.logical_or(angle2 < 0, angle2 > np.pi)
+    else:
+        adjust_mask = np.logical_or(angle2 < -np.pi / 2, angle2 > np.pi / 2)
+    angle1[adjust_mask] = angle1[adjust_mask] + np.pi
+    angle2[adjust_mask] = 2 * phi - angle2[adjust_mask]
+    angle3[adjust_mask] = angle3[adjust_mask] - np.pi
+
+    angles = np.column_stack((angle1, angle2, angle3))
+    angles[angles < -np.pi] += 2 * np.pi
+    angles[angles > np.pi] -= 2 * np.pi
+
+    # Step 8
+    # TODO: if any observability flags are poor, possibly raise a UserWarning?
```

Comment:
As I see it the degenerate angle might be only zero, because the troublesome value of the second angle will be right on the boundary of the valid range, so it won't be corrected. 

I see two approaches:

1. Issue a warning and explain that such case is possible in the docstring.
2. Not issue a warning but explain that such case is possible in the docstring.

I lean towards the second option.

Reply:
[Commenting here because the other one is being collapsed]
So I tested dcm equality for the degenerate case. The dcms are not equal upto 6 decimal places. Since it appears that the algorithm is not numerically stable for this case, will it be better to raise a Warning / Error? Or do we look for a better algorithm?

Reply:
I don't think there is a better algorithm. Now as you said it, a warning looks appropriate. I will investigate this point myself tonight.

Reply:
One particularly tricky case of gimbal lock I found was `Rotation.from_euler('yzx', [45, 90, 35], degrees=True)`. For this, `lambda = -90.0 degrees`, and all observability flags are `True`. Also note that the third angle returned was not `0`. The current implementation of `as_euler('yzx', degrees=True)` returns `[36.86989765, 89.99999879, 30.25643716]`.

It appears that we cannot detect all possible cases of gimbal lock by relying on the algorithm in the paper. I think that the paper's observability flags are not directly related to gimbal lock at all. This needs some discussion.

Reply:
I checked this case and I think I figured the problem. 

We have `dcm_transformed[:, 2, 2]` very close to -1 (should be -1 exactly) and then we have `np.arccos(dcm_transformed[:, 2, 2])` different from `pi` by something about `1e-8`. Speaking in exact numbers we definitely have the gimbal lock here. The problem occurs because you chose epsilon too small. By the way, why you chose it that way? Because of that we use "normal" formulas and get quite weird result.

I suggest to tune this eps such as to avoid such strange results for all sequences and when we set the second angle to create the gimbal lock situation (using degrees).

Reply:
I thought choosing `eps` to be at machine precision would be fine. I'll tune it.

Reply:
It is quite tricky and I'm not sure exactly how to analyze it. But your example showed that if we don't consider this case at all or choose eps too small then we will have bad results. Here we have `36.86989765 + 30.25643716 = 67.12633481` and the correct angles should add up to 80. In other words we got quite inaccurate dcm.

Reply:
Also I notices that for extrinsic angles we will have the first angle zero, not the third. Should we change it?

Reply:
Another thing I noticed: we probably should not apply `adjust_mask` for "gimbal lock cases". Try `Rotation.from_euler('YZX', [45, 90, 35], degrees=True)` to see what happens. The problem we have is that the second angle might be slightly off the desired boundary (less than epsilon), basically it means that this angle is very close to the critical value and we declare the "gimbal lock case" in our numerical algorithm. In other words we assume that this angle is equal to the gimbal lock angle exactly, it seems that we should not adjust it.

Reply:
So I put `eps = 1e-7`, since that was the smallest value that passes the test cases. I think it is fine since `assert_allclose` also has an absolute tolerance of 6 decimal places, but I want to know your opinion on this.

Reply:
> Also I notices that for extrinsic angles we will have the first angle zero, not the third. Should we change it?

I'm sorry change what exactly? Right now I'm detecting gimbal lock cases by the 'observability flag' mentioned in the paper.

Reply:
> Another thing I noticed: we probably should not apply adjust_mask for "gimbal lock cases". Try Rotation.from_euler('YZX', [45, 90, 35], degrees=True) to see what happens. The problem we have is that the second angle might be slightly off the desired boundary (less than epsilon), basically it means that this angle is very close to the critical value and we declare the "gimbal lock case" in our numerical algorithm. In other words we assume that this angle is equal to the gimbal lock angle exactly, it seems that we should not adjust it.

I see your point. For your example we get `[-100.,  89.99999879, -180.]` after adjustment. I'll remove the application of `adjust mask` for these cases. Our range restrictions may not apply for these cases but I'm not too worried since even in these cases our ranges will be off by at most `eps`.